---
title: "Hma Analysis"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Hma Analysis}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options:
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

```{r}
progressr::handlers("cli")
progressr::handlers(global = TRUE)
```

```{r}
path <- "D:/payle484/"
padlocout_master <- readr::read_rds(fs::path_join(c(path, "data/padlocout_master.rds")))
index <- readr::read_tsv("D:/payle484/refseq/index.txt")
```

```{r}
summary_system_count <- postpad::summarise_system_count(padlocout_master)
summary_system_overlap <- postpad::summarise_system_overlap(padlocout_master)
summary_protein_overlap <- postpad::summarise_protein_overlap(padlocout_master)

# HOW MANY HMA SYSTEMS?
dplyr::filter(summary_system_count, system == "hma")
```

```{r}
# Filter master padlocout table for Hma systems.
padlocout_hma <- dplyr::filter(padlocout_master, system == "hma")
# Join index information, so we know where to read GFF files from.
padlocout_hma_index <- dplyr::left_join(padlocout_hma, index, by = dplyr::join_by("genome.accession"))

# Extract GFF information of surrounding genes to 'build' loci.
future::plan(future.callr::callr, workers = 32)
hma_loci <- build_loci(padlocout_hma, index, path_db)

# Separate the attributes column and just keep the "Name" attribute (i.e. the 
# protein name; WP_...) and "locus_tag".
hma_loci_attr <- gff_separate_attributes(hma_loci, keep = c("Name", "locus_tag"))

make_seq_extract_index <- function(loci_tbl, index) {
  filtered <- dplyr::filter(loci_tbl, is.na(target.name))
  selected <- dplyr::select(filtered, genome.accession, Name)
  indexed <- dplyr::left_join(selected, index, by = dplyr::join_by("genome.accession"))
  out <- dplyr::select(indexed, Name, path_faa)
  out
}

seq_extract_index <- make_seq_extract_index(hma_loci_attr, index)[1:100,]

future::plan(future.callr::callr, workers = 16)
progressr::with_progress({
  seqs <- future_seq_extract(seqext_index, path_db)
})


# seqext <- function(id, file) {
#   command <- "awk"
#   args <- paste0("-v seq=", id, " -v RS='>' '$1 == seq {printf RS $0}' ", file)  
#   seq <- system2(command, args, stdout = TRUE)
#   seq
# }


# seqs <- purrr::map(
#   .x = cli::cli_progress_along(seq_along(1:nrow(proteins_index))),
#   .f = function(i) {
#     file <- fs::path_join(c(path_db, adjacent_proteins_index[["path_faa"]][i]))
#     id <- adjacent_proteins_index[["Name"]][i]
#     seq <- seqext(id, file)
#     seq
#   }
# )


# Read in the duplicates key generated by seqkit.
duplicates_key <- read_seqkit_duplicates("D:/payle484/data/duplicates.txt")

# Read in the cluster key generated by mmseqs.
cluster_key <- read_mmseqs_cluster_tsv("D:/payle484/mmseqs-cluster/results/clusters/hma_adjacent/hma_adjacent.tsv")

# Unnest the duplicates key so each row is a distinct pair of representative 
# and member.
duplicates_key_unnested <- tidyr::unnest(duplicates_key, dup_member)
duplicates_key_distinct <- dplyr::distinct(duplicates_key_unnested)

# Assign the representative names to duplicate proteins that got discarded by 
# seqkit before clustering with mmseqs.
hma_loci_dups <- dplyr::left_join(hma_loci_attr, duplicates_key_distinct, by = dplyr::join_by("Name" == "dup_member"))
hma_loci_dups_filled <- dplyr::mutate(hma_loci_dups, dup_representative = ifelse(is.na(dup_representative), Name, dup_representative))

# Assign the cluster names.
hma_loci_clusters <- dplyr::left_join(hma_loci_dups_filled, cluster_key, by = dplyr::join_by("dup_representative" == "clu_member"))
# Fill the cluster name column with the first non-NA value out of the 
# cluster names, protein names, and locus tags respectively, effectively
# assigning any missing cluster names with the protein name if they're an Hma 
# protein or the locus tag if they're an adjacent protein
hma_loci_clusters_filled <- dplyr::mutate(hma_loci_clusters, clu_representative = dplyr::coalesce(clu_representative, protein.name, locus_tag))

check <- hma_loci_clusters %>%
  # Is the protein an 'adjacent' protein (i.e. in the locus and not the primary system)?
  dplyr::mutate(is_adjacent = ifelse(is.na(target.name), TRUE, FALSE)) %>%
  dplyr::filter(is_adjacent == TRUE) %>%
  # Does the protein have the 'Name' attribute in the respective GFF file? 
  # (pseudogenes won't unless we've put the substituted name in the GFF file).
  dplyr::mutate(has_name = ifelse(is.na(Name), FALSE, TRUE)) %>%
  # Does the protein have a cluster representative ?
  # (pseudogenes won't unless we've put the substituted sequence in the FASTA file)
  dplyr::mutate(has_clu_representative = ifelse(is.na(clu_representative), FALSE, TRUE))

check_summary <- 
  dplyr::summarise(check, n = dplyr::n(), .by = c(is_adjacent, has_name, has_clu_representative))

check_filter <- 
  dplyr::filter(check, has_name == FALSE | has_clu_representative == FALSE) %>%
  dplyr::select(genome.accession, seqid, start, end, strand, dup_representative, clu_representative, path_gff, is_adjacent, has_name, has_clu_representative) %>%
  dplyr::left_join(hma_loci %>% dplyr::select(genome.accession, seqid, start, end, strand, path_gff, attributes),
                   by = dplyr::join_by(genome.accession, seqid, start, end, strand, path_gff)) %>%
  gff_separate_attributes()

check_filter %>% dplyr::distinct(genome.accession) %>% nrow()

loci_unique <- loci_clu %>% 
  dplyr::ungroup() %>%
  dplyr::mutate(cluster = ifelse(is.na(cluster), protein.name, cluster)) %>%
  dplyr::group_by(genome.accession, seqid, locus_name) %>% 
  dplyr::arrange(cluster) %>% # Just arrange loci alphabetically by cluster
  dplyr::mutate(
    locus = paste0(cluster, collapse = ">>")
  ) %>%
  dplyr::select(genome.accession, seqid, locus_name, locus) %>%
  dplyr::distinct() %>%
  dplyr::group_by(locus) %>%
  dplyr::slice_head() %>%
  dplyr::ungroup() %>%
  dplyr::select(-locus) %>%
  dplyr::mutate(keep = TRUE)

# HOW MANY DISTINCT HMA SYSTEMS?
loci_named <- dplyr::mutate(hma_loci_clusters_filled, locus_name = paste0(genome.accession, "_", seqid, "_", system, "_", system.number))
loci_summarised <- dplyr::mutate(loci_named, )

loci_tbl_mutate_locus_structure <- function(loci_tbl) {
  arranged <- dplyr::arrange(loci_tbl, clu_representative)
  mutated <- dplyr::mutate(
    arranged, 
    locus_structure = paste0(clu_representative, collapse = ">>"), 
    .by = c(genome.accession, seqid, system, system.number))
  out <- dplyr::arrange(mutated, genome.accession, seqid, system, system.number, relative.position)
  out
}

loci_tbl <- hma_loci_clusters_filled

tmp_nested <- tidyr::nest(loci_nested, .by = loci) 
tmp_nested_count <- tmp_nested %>% dplyr::mutate(n = purrr::map_dbl(data, ~ nrow(.x))) %>% dplyr::arrange(desc(n))

loci_tbl_nest_structure <- function(loci_tbl) {
  loci_tbl_w_structure <- loci_tbl_mutate_locus_structure(loci_tbl)
  loci_nested <- tidyr::nest(loci_tbl_w_structure, .by = c(genome.accession, seqid, system, system.number, locus_structure))
  loci_nested_count <- dplyr::mutate(loci_nested, locus_strucutre_n = dplyr::n(), .by = locus_structure)
  loci_stucture_nested <- tidyr::nest(loci_nested_count, .by = c(locus_structure, locus_strucutre_n))
  out <- dplyr::arrange(loci_stucture_nested, desc(locus_strucutre_n))
}

loci_tbl_nested <- loci_tbl_nest_structure(loci_tbl)

loci_tbl_nested_sample <- function(loci_tbl_nested) {
  sample <- purrr::map(loci_tbl_nested$data, ~ dplyr::slice_sample(.x))
  loci_tbl_nested$data <- sample
  loci_tbl_nested
}

set.seed(14561983)
loci_sample_nested <- loci_tbl_nested_sample(loci_tbl_nested)

loci_sample <- tidyr::unnest(loci_sample_nested, data)
sampled_genomes <- dplyr::pull(loci_sample, genome.accession)

padlocout_master_sampled <- dplyr::filter(padlocout_master, genome.accession %in% sampled_genomes)


dplyr::pull(tidyr::unnest(loci_sample, data), genome.accession)

loci_sample_genomes <- tidyr::pluck()

loci_unq_sample <- purrr::map(tmp_nested_count$data, ~ dplyr::slice_sample(.x))

tmp_nested$data

loci_tbl <- hma_loci

loci_distinct <- function(loci_tbl, cluster_key) {
  loci_tbl_attr <- gff_separate_attributes(hma_loci, keep = "Name")
  loci_tbl_clu <- dplyr::left_join(hma_loci_attr, cluster_key, by = dplyr::join_by("Name" == "member"))
}

tmp <- hma_loci_clusters_filled %>% 
  dplyr::ungroup() %>%
  dplyr::mutate(locus_name = paste0(genome.accession, "_", seqid, "_", system, "_", system.number)) %>%
  dplyr::group_by(locus_name) %>% 
  # dplyr::arrange(relative_relative_position) %>%
  dplyr::arrange(clu_representative) %>% # Just arrange loci alphabetically by cluster representative
  dplyr::mutate(locus = paste0(clu_representative, collapse = ">>")) %>%
  dplyr::distinct(locus_name, locus) %>%
  dplyr::group_by(locus) %>%
  dplyr::summarise(n = dplyr::n()) %>%
  dplyr::arrange(desc(n)) %>%
  dplyr::mutate(
    rank = dplyr::row_number(),
    prop = n / sum(n)
  )

tmp %>% ggplot2::ggplot() +
  ggplot2::geom_col(ggplot2::aes(x = rank, y = n))

tmp2 <- loci_clu %>% 
  dplyr::ungroup() %>%
  dplyr::mutate(cluster = ifelse(is.na(cluster), protein.name, cluster)) %>%
  dplyr::group_by(genome.accession, seqid, locus_name) %>% 
  dplyr::arrange(cluster) %>% # Just arrange loci alphabetically by cluster
  dplyr::mutate(
    locus = paste0(cluster, collapse = ">>")
  ) %>%
  dplyr::arrange(genome.accession, seqid, locus_name, relative.position) %>%
  dplyr::ungroup()

tmp2 %>% 
  dplyr::filter(locus == tmp[[1,1]]) %>% 
  dplyr::group_by(genome.accession, seqid, locus_name) %>%
  dplyr::mutate(locus = paste0(cluster, collapse = "__")) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(locus) %>%
  dplyr::summarise(n = dplyr::n())



embedded_hma <- purrr::map(
  .x = cli::cli_progress_along(padlocout_list, name = "Finding loci"),
  .f = function(i) {
    gff_path <- fs::path_join(c(path_db, unique(padlocout_list[[i]]$path_gff)))
    gff <- read_gff(gff_path)
    gff_cds <- dplyr::filter(gff, type == "CDS")
    gff_cds_pos <- assign_position(gff_cds)
    gff_joined <- gff_join_padlocout(gff_cds_pos, padlocout_list[[i]])
    gff_joined_filtered <- gff_filter_locus(padlocout_list[[i]], gff_joined, 0)
    gff_joined_filtered
  }
)

names(embedded_hma) <- names(padlocout_master_hma_list)

embedded_hma_combined <- loci_combine(embedded_hma)

embedded_hma_combined_unq <- embedded_hma_combined %>%
  dplyr::left_join(loci_unique, by = dplyr::join_by(genome.accession, seqid, locus_name)) %>%
  dplyr::filter(keep == TRUE)

embedded_hma_combined_unq %>% 
  dplyr::group_by(genome.accession, seqid, locus_name) %>%
  dplyr::mutate(has_embedded = ifelse(any(is.na(protein.name)), TRUE, FALSE)) %>%
  dplyr::select(locus_name, has_embedded) %>% 
  dplyr::distinct() %>%
  dplyr::group_by(has_embedded) %>%
  dplyr::summarise(n = dplyr::n())


```



