---
title: "sandbox"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{sandbox}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options:
  chunk_output_type: console
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = FALSE
)
```

```{r}
# BENCHMARKING SEQUENCE EXTRACTION
# ------------------------------------------------------------------------------

file <- "~/sync/personal/projects/padloc/test/GCF_001688665.2.faa"

ids_0001 <- sample(ids, 1)
ids_0010 <- sample(ids, 10)
ids_0100 <- sample(ids, 100)
ids_1000 <- sample(ids, 1000)

microbenchmark::microbenchmark(ids_0001_out <- purrr::map(.x = ids_0001, .f = function(x) seqext(x, file)), times = 100)
# Unit: milliseconds
#       min       lq     mean   median       uq      max neval
#  128.1251 129.8642 137.9504 132.3454 140.8534 184.6368   100
microbenchmark::microbenchmark(ids_0010_out <- purrr::map(.x = ids_0010, .f = function(x) seqext(x, file)), times = 100)
# Unit: seconds
#       min       lq     mean   median      uq      max neval
#  1.297341 1.322547 1.395876 1.349584 1.41204 2.173802   100
microbenchmark::microbenchmark(ids_0100_out <- purrr::map(.x = ids_0100, .f = function(x) seqext(x, file)), times = 10)
# Unit: seconds
#       min       lq     mean  median       uq      max neval
# 13.21945 13.25569 13.47494 13.4514 13.63052 13.95471     10
microbenchmark::microbenchmark(ids_1000_out <- purrr::map(.x = ids_1000, .f = function(x) seqext(x, file)), times = 1)
# Unit: seconds
#       min       lq     mean   median       uq      max neval
#  143.8383 143.8383 143.8383 143.8383 143.8383 143.8383     1
```

```{r}


# GROUP BY MOLECULAR UNITS, TESTING


df <- tibble::tibble(
  molecule = c("mol_01", "mol_01", "mol_01", "mol_01", "mol_01", "mol_02", "mol_02", "mol_02", "mol_02", "mol_02"),
  system   = c("sys_01", "sys_01", "sys_02", "sys_02", "sys_03", "sys_01", "sys_01", "sys_04", "sys_04", "sys_05"),
  gene     = c("gen_01", "gen_02", "gen_01", "gen_01", "gen_02", "gen_01", "gen_02", "gen_01", "gen_02", "gen_01"),
  start    = c(       1,      121,      601,      841,     1801,        1,      121,      841,      961,     1081),
  end      = c(     100,      220,      700,      940,     1900,      100,      220,      940,     1060,     1180),
  position = c(       1,        2,        5,        7,       15,        1,        2,        7,        8,        9)
)

# First add columns for the minimum and maximum positions for each system, so 
# systems already grouped don't get lost.
df_grouped <- add_group_limits(df, position, position, .by = c("molecule", "system"))
df_grouped <- add_group_limits(df_grouped, start, end, .by = c("molecule", "system"))

# Not necessary, but simplifies the data frame.
df_nested <- tidyr::nest(df_grouped, .by = c("molecule", "system", "min_system_position", "max_system_position", "min_system_start", "max_system_end"))

group_by_distance(df_nested, min_system_position, max_system_position, n = 1, name = "island", .by = "molecule")
group_by_distance(df_nested, min_system_start, max_system_end, n = 500, name = "island", .by = "molecule")


```

```{r}
# Create master padlocout

path <- "D:/payle484/"

DRT_combined           <- read_padlocout_master(fs::path_join(c(path, "DRT_combined_padloc.csv")))
HEC_combined           <- read_padlocout_master(fs::path_join(c(path, "HEC_combined_padloc.csv")))
MG1_combined           <- read_padlocout_master(fs::path_join(c(path, "MG1_combined_padloc.csv")))
new_combined           <- read_padlocout_master(fs::path_join(c(path, "new_combined_padloc.csv")))
SG1_combined           <- read_padlocout_master(fs::path_join(c(path, "SG1_combined_padloc.csv")))
SigPro_combined        <- read_padlocout_master(fs::path_join(c(path, "SigPro_combined_padloc.csv")))
vassallo_2022_combined <- read_padlocout_master(fs::path_join(c(path, "vassallo_2022_combined_padloc.csv")))
CRISPR_combined        <- read_padlocout_master(fs::path_join(c(path, "CRISPR_combined_padloc.csv")))
DG1_combined           <- read_padlocout_master(fs::path_join(c(path, "DG1_combined_padloc.csv")))
DMS_combined           <- read_padlocout_master(fs::path_join(c(path, "DMS_combined_padloc.csv")))

padlocout_master <- dplyr::bind_rows(
  DRT_combined,
  HEC_combined,
  MG1_combined,
  new_combined,
  SG1_combined,
  SigPro_combined,
  vassallo_2022_combined,
  CRISPR_combined,
  DG1_combined,
  DMS_combined
)

padlocout_master_list <- padlocout_uncombine(padlocout_master)

# readr::write_rds(padlocout_master, fs::path_join(c(path, "padlocout_master.rds")))
# readr::write_rds(padlocout_master_list, fs::path_join(c(path, "padlocout_master_list.rds")))


padlocout_master_index <- dplyr::left_join(padlocout_master, index, by = dplyr::join_by("genome.accession"))

padlocout_master_index_00 <- padlocout_master_index %>% dplyr::filter(chunk == "00")

padlocout_master_index_00_list <- padlocout_uncombine(padlocout_master_index_00)

path_db <- "D:/payle484/refseq/"
tmp <- purrr::map(
  .x = padlocout_master_index_00_list,
  .f = function(x) {
    gff_path <- fs::path_join(c(path_db, unique(x$path_gff)))
    gff <- read_gff(gff_path)
    gff_cds <- dplyr::filter(gff, type == "CDS")
    gff_cds_pos <- assign_position(gff_cds)
    gff_joined <- gff_join_padlocout(gff_cds_pos, x)
    gff_joined_filtered <- gff_filter_locus(x, gff_joined, 3)
    gff_joined_filtered
  }
)



padlocout_list <- padlocout_master_hma_list

loci_hma <- purrr::map(
  .x = cli::cli_progress_along(padlocout_list, name = "Finding loci"),
  .f = function(i) {
    gff_path <- fs::path_join(c(path_db, unique(padlocout_list[[i]]$path_gff)))
    gff <- read_gff(gff_path)
    gff_cds <- dplyr::filter(gff, type == "CDS")
    gff_cds_pos <- assign_position(gff_cds)
    gff_joined <- gff_join_padlocout(gff_cds_pos, padlocout_list[[i]])
    gff_joined_filtered <- gff_filter_locus(padlocout_list[[i]], gff_joined, 3)
    gff_joined_filtered
  }
)

#### WORKING ####





names(loci_hma) <- names(padlocout_master_hma_list)

loci_list <- loci_hma

# loci_combine <- function(loci_list) {
#   loci_unlisted <- purrr::map(
#     .x = cli::cli_progress_along(loci_list, name = "Unlisting loci"),
#     .f = function(i) {
#       purrr::map(
#         .x = loci_list[[i]],
#         .f = function(x) dplyr::mutate(x, locus_name = names(x))
#       )
#       unlisted <- dplyr::bind_rows(loci_list[[i]])
#       unlisted
#     }
#   )
#   names(loci_unlisted) <- names(loci_list)
#   combined <- padlocout_combine(loci_unlisted)
# }



loci_combined <- loci_combine(loci_hma)





compact_loci <- function(loci_list) {

}





combine_gff <- function(gff_list) {
  to_combine <- purrr::map(
    .x = cli::cli_progress_along(gff_list, name = "Combining GFF files"),
    .f = function(i) {
      # Check if there's a column for genome accession, and if not make one
      if (! "genome.accession" %in% names(gff_list[[i]])) {
        file_name <- names(gff_list[i])
        genome_accession <- stringr::str_remove(file_name, "_genomic.gff")
        x <- dplyr::mutate(gff_list[[i]], "genome.accession" = genome_accession)
        x <- dplyr::select(x, "genome.accession", dplyr::everything())
        x
      } else {
        gff_list[[i]]
      }
    }
  )
  combined <- dplyr::bind_rows(to_combine)
  combined
}




tmp <- read_fasta_file("D:/payle484/refseq/faa/00/GCF_000003955.1_ASM395v1_protein.faa")
tmp[names(tmp) %in% c("WP_000278981.1", "WP_000262736.1", "WP_000229708.1")]


seqs <- purrr::map(
  .x = cli::cli_progress_along(seq_along(1:nrow(proteins_index))),
  .f = function(i) {
    file <- fs::path_join(c(path_db, adjacent_proteins_index[["path_faa"]][i]))
    id <- adjacent_proteins_index[["Name"]][i]
    seq <- seqext(id, file)
    seq
  }
)


proteins_index <- adjacent_proteins_index



tmp <- furrr::future_map2(
  .x = proteins_index[["path_faa"]],
  .y = proteins_index[["Name"]],
  .f = ~ {
    file <- fs::path_join(c(path_db, .x))
    seq <- seqext(.y, file)
    seq
  }
)

proteins_index <- adjacent_proteins_index[1:10000,]



future::plan(future.callr::callr, workers = 16)

tictoc::tic()
progressr::with_progress(
  {tmp2 <- future_seqext(proteins_index, path_db)}
  ) 
tictoc::toc()

future::plan(future::sequential)

# 32 workers, callr, 15795 proteins = 609.14 sec

tmp_unlisted <- unlist(tmp)
readr::write_lines(tmp_unlisted, "D:/payle484/data/hma_adjacent_proteins.faa", )


clukey <- readr::read_tsv("D:/payle484/mmseqs/hma_adjacent_clu_cluster.tsv", col_names = c("cluster", "member"), col_types = c("c", "c")) %>% dplyr::distinct(member, .keep_all = TRUE)

loci_nameattr <- gff_separate_attributes(loci_combined, keep = "Name")

loci_clu <- dplyr::left_join(loci_nameattr, clukey, by = dplyr::join_by("Name" == "member"), multiple = "all")

tmp <- loci_clu %>% 
  dplyr::ungroup() %>%
  # dplyr::mutate(relative_relative_position = ifelse(strand == "+", relative.position, -relative.position)) #%>% 
  dplyr::mutate(cluster = ifelse(is.na(cluster), protein.name, cluster)) %>%
  dplyr::group_by(genome.accession, seqid, locus_name) %>% 
  # dplyr::arrange(relative_relative_position) %>%
  dplyr::arrange(cluster) %>% # Just arrange loci alphabetically by cluster
  dplyr::mutate(
    locus = paste0(cluster, collapse = ">>")
  ) %>%
  dplyr::distinct(genome.accession, seqid, locus_name, locus) %>%
  dplyr::group_by(locus) %>%
  dplyr::summarise(n = dplyr::n()) %>%
  dplyr::arrange(desc(n))

tmp2 <- loci_clu %>% 
  dplyr::ungroup() %>%
  dplyr::mutate(cluster = ifelse(is.na(cluster), protein.name, cluster)) %>%
  dplyr::group_by(genome.accession, seqid, locus_name) %>% 
  dplyr::arrange(cluster) %>% # Just arrange loci alphabetically by cluster
  dplyr::mutate(
    locus = paste0(cluster, collapse = ">>")
  ) %>%
  dplyr::arrange(genome.accession, seqid, locus_name, relative.position) %>%
  dplyr::ungroup()

tmp2 %>% 
  dplyr::filter(locus == tmp[[1,1]]) %>% 
  dplyr::group_by(genome.accession, seqid, locus_name) %>%
  dplyr::mutate(locus = paste0(cluster, collapse = "__")) %>%
  dplyr::ungroup() %>%
  dplyr::group_by(locus) %>%
  dplyr::summarise(n = dplyr::n())

loci_unique <- loci_clu %>% 
  dplyr::ungroup() %>%
  dplyr::mutate(cluster = ifelse(is.na(cluster), protein.name, cluster)) %>%
  dplyr::group_by(genome.accession, seqid, locus_name) %>% 
  dplyr::arrange(cluster) %>% # Just arrange loci alphabetically by cluster
  dplyr::mutate(
    locus = paste0(cluster, collapse = ">>")
  ) %>%
  dplyr::select(genome.accession, seqid, locus_name, locus) %>%
  dplyr::distinct() %>%
  dplyr::group_by(locus) %>%
  dplyr::slice_head() %>%
  dplyr::ungroup() %>%
  dplyr::select(-locus) %>%
  dplyr::mutate(keep = TRUE)

embedded_hma <- purrr::map(
  .x = cli::cli_progress_along(padlocout_list, name = "Finding loci"),
  .f = function(i) {
    gff_path <- fs::path_join(c(path_db, unique(padlocout_list[[i]]$path_gff)))
    gff <- read_gff(gff_path)
    gff_cds <- dplyr::filter(gff, type == "CDS")
    gff_cds_pos <- assign_position(gff_cds)
    gff_joined <- gff_join_padlocout(gff_cds_pos, padlocout_list[[i]])
    gff_joined_filtered <- gff_filter_locus(padlocout_list[[i]], gff_joined, 0)
    gff_joined_filtered
  }
)

names(embedded_hma) <- names(padlocout_master_hma_list)

embedded_hma_combined <- loci_combine(embedded_hma)

embedded_hma_combined_unq <- embedded_hma_combined %>%
  dplyr::left_join(loci_unique, by = dplyr::join_by(genome.accession, seqid, locus_name)) %>%
  dplyr::filter(keep == TRUE)

embedded_hma_combined_unq %>% 
  dplyr::group_by(genome.accession, seqid, locus_name) %>%
  dplyr::mutate(has_embedded = ifelse(any(is.na(protein.name)), TRUE, FALSE)) %>%
  dplyr::select(locus_name, has_embedded) %>% 
  dplyr::distinct() %>%
  dplyr::group_by(has_embedded) %>%
  dplyr::summarise(n = dplyr::n())







```

```{r}
# REASSEMBLE A DIVIDED PADLOCDB

path <- "D:/payle484/padlocdb-final/padloc_data"

read_split_hmm_meta <- function(path) {
  dirs <- fs::dir_ls(path)
  hmm_meta_paths <- purrr::map_vec(dirs, ~ fs::path_join(c(.x, "hmm_meta.txt")))
  hmm_meta_list <- purrr::map(hmm_meta_paths, ~ padlocdev:::read_hmm_meta(.x))
  hmm_meta_list
}

hmm_meta_combine <- function(hmm_meta_list) {
  hmm_meta_bound <- dplyr::bind_rows(hmm_meta_list)
  hmm_meta_distinct <- dplyr::distinct(hmm_meta_bound, hmm.acc, .keep_all = TRUE)
  hmm_meta_arranged <- dplyr::arrange(hmm_meta_distinct, hmm.acc)
  hmm_meta_arranged
}

hmm_meta_list <- read_split_hmm_meta(path)
hmm_meta <- hmm_meta_combine(hmm_meta_list)
hmm_meta <- padlocdev::expand_hmm_meta(hmm_meta)

read_split_sys_meta <- function(path) {
  dirs <- fs::dir_ls(path)
  sys_meta_paths <- purrr::map_vec(dirs, ~ fs::path_join(c(.x, "sys_meta.txt")))
  sys_meta_list <- purrr::map(sys_meta_paths, ~ padlocdev:::read_sys_meta(.x))
  sys_meta_list
}

sys_meta_combine <- function(sys_meta_list) {
  sys_meta_bound <- dplyr::bind_rows(sys_meta_list)
  sys_meta_distinct <- dplyr::distinct(sys_meta_bound, yaml.name, .keep_all = TRUE)
  sys_meta_arranged <- dplyr::arrange(sys_meta_distinct, yaml.name)
  sys_meta_arranged
}

sys_meta_list <- read_split_sys_meta(path)
sys_meta <- sys_meta_combine(sys_meta_list)

read_split_padloc_model <- function(path) {
  dirs <- fs::dir_ls(path)
  sys_dirs <- purrr::map_vec(dirs, ~ fs::path_join(c(.x, "sys")))
  models_list <- purrr::map(sys_dirs, ~ padlocdev::multi_read_padloc_model(.x))
  names(models_list) <- NULL
  models <- unlist(models_list, recursive = FALSE)
  models
}

models <- read_split_padloc_model(path)

hmmer3db_separate <- function(path, out = NULL, remove = FALSE) {

  if ( checkmate::check_null(out) ) out <- fs::path_dir(path)
  
  # read in the HMMER3 profile HMM database
  hmmer3db <- readr::read_lines(path, progress = FALSE)
  
  # extract the individual HMMs and their accession headers
  hmm_heads <- which(stringr::str_detect(hmmer3db, "^HMMER3/f"))
  hmm_tails <- which(stringr::str_detect(hmmer3db, "^//"))
  hmm_accs  <- which(stringr::str_detect(hmmer3db, "^ACC"))
  hmm_accs  <- stringr::str_extract(hmmer3db[hmm_accs], "(?<=ACC   ).*")
  
  hmm_count <- length(hmm_heads)
  
  hmms <- purrr::map2(hmm_heads, hmm_tails, ~ paste0(hmmer3db[.x:.y], collapse = "\n"))
  
  purrr::iwalk(
    hmms, 
    function(x, idx) {
      readr::write_file(x, fs::path_join(c(out, paste0(hmm_accs[idx], ".hmm"))))
    }
  )
  
  if ( remove == TRUE ) fs::file_delete(path)

}

hmmer3db_paths <- 
  fs::dir_ls("D:/payle484/padlocdb-final/padloc_data/") %>%
  purrr::map_vec(~ fs::path_join(c(.x, "hmm/padlocdb.hmm")))

purrr::walk(hmmer3db_paths, ~ hmmer3db_separate(.x, remove = TRUE), .progress = TRUE)

read_split_hmm_header <- function(path) {
  dirs <- fs::dir_ls(path)
  hmm_dirs <- purrr::map_vec(dirs, ~ fs::path_join(c(.x, "hmm")))
  hmms_list <- purrr::map(hmm_dirs, ~ padlocdev::multi_read_hmm_header(.x), .progress = TRUE)
  names(hmms_list) <- NULL
  hmms <- unlist(hmms_list, recursive = FALSE)
  hmms_unique <- hmms[!duplicated(paste(hmms, names(hmms)))]
  hmms_unique
}

hmms <- read_split_hmm_header(path)

# Verify that HMM file names match their accession.
padlocdev::verify_hmm_names(hmms)
# Compare the profile HMMs with the HMM metadata table.
padlocdev::compare_hmm_files_hmm_meta(hmms, hmm_meta)
# Compare the system models with the system metadata table.
padlocdev::compare_sys_files_sys_meta(models, sys_meta)
# Expand the HMM metadata table to account for HMMs with multiple name 
# assignments, and expand the system models to account for groups of genes.
hmm_meta_expanded <- padlocdev::expand_hmm_meta(hmm_meta)
models_expanded   <- padlocdev::expand_gene_groups_all(models, hmm_meta_expanded)
# Compare the system models with the HMM metadata table.
padlocdev::compare_sys_files_hmm_meta(models_expanded, hmm_meta_expanded)
# Generate a validity report for the system models. Inspect the validity report 
# to confirm that all models are valid. Update the models where necessary, then 
# re-read and re-validate to confirm.
validity_report <- padlocdev::report_padloc_model_validity(models_expanded)
padlocdev::why_invalid(validity_report)

# Write-out
fs::dir_create("D:/payle484/shuffling/padloc-db/sys")

hmm_meta_out <- dplyr::mutate(hmm_meta, dplyr::across(dplyr::everything(), as.character))
readr::write_tsv(hmm_meta_out, "D:/payle484/shuffling/padloc-db/hmm_meta.txt")

readr::write_tsv(sys_meta, "D:/payle484/shuffling/padloc-db/sys_meta.txt")

padlocdev::multi_write_padloc_model(models, "D:/payle484/shuffling/padloc-db/sys/")
```
