#' Combine a list of padlocout tables
#' @param padlocout_list A list of padlocout tables (as read in by
#' [multi_read_padlocout()]).
#' @return A combined padlocout table, introducing a new row `genome_accession`,
#' which is derived from the names of the padlocout tables in `padlocout_list`.
#' @export
combine_padlocout <- function(padlocout_list) {
  to_combine <- purrr::map(
    .x = cli::cli_progress_along(padlocout_list, name = "Combining PADLOC output"),
    .f = function(i) {
      # Check if there's a column for genome accession, and if not make one
      if (! "genome.accession" %in% names(padlocout_list[[i]])) {
        file_name <- names(padlocout_list[i])
        genome_accession <- stringr::str_remove(file_name, "_protein_padloc.csv")
        x <- dplyr::mutate(padlocout_list[[i]], "genome.accession" = genome_accession)
        x <- dplyr::select(x, "genome.accession", dplyr::everything())
        x
      } else {
        padlocout_list[[i]]
      }
    }
  )
  combined <- dplyr::bind_rows(to_combine)
  combined
}

#' Summarise how many of each system were identified
#' @param padlocout A padlocout table (as read in by [read_padlocout()]) or a
#' combined padlocout table (as generated by [combine_padlocout()])
#' @return A [tibble::tibble()] with columns `system` and `n`, where there is a
#' single row for each system summarising the number of times a system was
#' identified in the padlocout table.
#' @export
summarise_system_count <- function(padlocout) {
  grouped <- dplyr::group_by(padlocout, seqid, system)
  unique <- dplyr::distinct(grouped, system.number)
  grouped <- dplyr::group_by(unique, system)
  count <- dplyr::summarise(grouped, n = dplyr::n())
  count
}

filter_gff <- function() {
  NULL
}

calculate_separation <- function(padlocout) {
  a <- dplyr::group_by(padlocout, seqid, system, system.number)
  b <- dplyr::mutate(a, max = max(relative.position), min = min(relative.position), genes = dplyr::n())
  c <- dplyr::mutate(b, separation = max + 1 - min - genes)
  d <- dplyr::select(c, -c(max, min, genes))
  d
}

assign_position <- function(gff) {
  gff_arranged <- dplyr::arrange(gff, seqid, start)
  gff_grouped <- dplyr::group_by(gff_arranged, seqid, type)
  out <- dplyr::mutate(gff_grouped, relative.position = dplyr::row_number())
}



